### ISSUE TITLE
What is soft margin? If we use hard margin the distance difference should be less than hard margin for the same objects?

### ISSUE THREAD

**Issue by voqtuyen** [VisualComputingInstitute/triplet-reid#62]
If we use hard margin the distance difference should be less than hard margin for the same objects? Is it correct?
I see that after retraining the code with the provided configs and hard-margin =0.5 but when infer the model with 2 images of the same object, the distance is still very large compared to hard-magin value

Thanks

**Comment by Pandoro** [comment]
As I already mentioned in #51, the distance between images of the same object are not directly constrained at all by the triplet loss. The margin is just added on top of the distance between positives samples and that is used to determine the minimal distance to the negative samples.

I'm quite sure that after training this margin constraint will not be violated by most of the training set, but as also mentioned before, the embeddings are more noisy on the test set and will thus likely result in both negatives and positives being quite close to some anchors.

You could add another loss term that tries to push together positives without any limit, essentially trying to collapse them to a single point. This might potentially work better, but we never experimented with this much. It reminds me of the contrastive loss by Hadsell et al., in which case the margin likely has a more important and definite meaning, and of [this paper](https://arxiv.org/abs/1603.02139), which you should take with a big spoon of salt though, since they do not achieve anywhere near state-of-the-art results.

**Comment by voqtuyen** [comment]
Thanks for your info, but it seems for me that the trained model produces large distance (> 7*margin), but not smaller than margin even on the training set
